x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java

```diff
@@ -347,5 +350,30 @@ public PersistentTasksExecutor<?> getPersistentTasksExecutor(Client client, Thre
                 SecurityMigrations.MIGRATIONS_BY_VERSION
             );
         }
+
+        private int getNumberOfAttempts(ProjectId project) {
+            var retryCount = taskSubmissionAttemptCounter.get(project);
+            if (retryCount == null) {
+                return 0;
+            }
+            return retryCount.get();
+        }
+
+        private void projectDeleted(ProjectId projectId) {
+            resetNumberOfAttempts(projectId);
+        }
+
+        private void resetNumberOfAttempts(ProjectId project) {
+            taskSubmissionAttemptCounter.remove(project);
+        }
+
+        private void incrementAttemptCount(ProjectId project) {
+            taskSubmissionAttemptCounter.computeIfAbsent(project, ignore -> new AtomicInteger(0)).incrementAndGet();
+        }
+
+        private void decrementAttemptCount(ProjectId project) {
+            taskSubmissionAttemptCounter.computeIfAbsent(project, ignore -> new AtomicInteger(0)).decrementAndGet();
```

Comment:
This can potentially lead to negative value as well as recreating the entry after its deletion. I think it's better to be a noop if the counter is not found for the project. The increment method seems fine since it is invoked inside the applier thread.